kernel/meson.build

```diff
@@ -795,17 +796,42 @@ base_kops = [
         }},
+        # TODO(rg): These can be cleaned up, c -> -UDOUBLE -DCOMPLEX anyway
         'c': {'exts': {
             '_iutcopy': {'addl': ['-UDOUBLE', '-DCOMPLEX', '-UOUTER', '-ULOWER']},
             '_iltcopy': {'addl': ['-UDOUBLE', '-DCOMPLEX', '-UOUTER', '-DLOWER']},
             '_outcopy': {'addl': ['-UDOUBLE', '-DCOMPLEX', '-DOUTER', '-ULOWER']},
             '_oltcopy': {'addl': ['-UDOUBLE', '-DCOMPLEX', '-DOUTER', '-DLOWER']},
+            '3m_iucopyb': {'addl': ['-UDOUBLE', '-DCOMPLEX', '-UUSE_ALPHA']},
+            '3m_oucopyb': {'addl': ['-UDOUBLE', '-DCOMPLEX', '-DUSE_ALPHA']},
+            '3m_iucopyr': {'addl': ['-UDOUBLE', '-DCOMPLEX', '-UUSE_ALPHA', '-DREAL_ONLY']},
+            '3m_oucopyr': {'addl': ['-UDOUBLE', '-DCOMPLEX', '-DUSE_ALPHA', '-DREAL_ONLY']},
+            '3m_iucopyi': {'addl': ['-UDOUBLE', '-DCOMPLEX', '-UUSE_ALPHA', '-DIMAGE_ONLY']},
+            '3m_oucopyi': {'addl': ['-UDOUBLE', '-DCOMPLEX', '-DUSE_ALPHA', '-DIMAGE_ONLY']},
+            '3m_ilcopyb': {'addl': ['-UDOUBLE', '-DCOMPLEX', '-UUSE_ALPHA']},
+            '3m_olcopyb': {'addl': ['-UDOUBLE', '-DCOMPLEX', '-DUSE_ALPHA']},
+            '3m_ilcopyr': {'addl': ['-UDOUBLE', '-DCOMPLEX', '-UUSE_ALPHA', '-DREAL_ONLY']},
+            '3m_olcopyr': {'addl': ['-UDOUBLE', '-DCOMPLEX', '-DUSE_ALPHA', '-DREAL_ONLY']},
+            '3m_ilcopyi': {'addl': ['-UDOUBLE', '-DCOMPLEX', '-UUSE_ALPHA', '-DIMAGE_ONLY']},
+            '3m_olcopyi': {'addl': ['-UDOUBLE', '-DCOMPLEX', '-DUSE_ALPHA', '-DIMAGE_ONLY']},
         }},
+        # TODO(rg): These can be cleaned up, z -> -DDOUBLE -DCOMPLEX anyway
         'z': {'exts': {
             '_iutcopy': {'addl': ['-DDOUBLE', '-DCOMPLEX', '-UOUTER', '-ULOWER']},
             '_iltcopy': {'addl': ['-DDOUBLE', '-DCOMPLEX', '-UOUTER', '-DLOWER']},
             '_outcopy': {'addl': ['-DDOUBLE', '-DCOMPLEX', '-DOUTER', '-ULOWER']},
             '_oltcopy': {'addl': ['-DDOUBLE', '-DCOMPLEX', '-DOUTER', '-DLOWER']},
+            '3m_iucopyb': {'addl': ['-DDOUBLE', '-DCOMPLEX', '-UUSE_ALPHA']},
+            '3m_oucopyb': {'addl': ['-DDOUBLE', '-DCOMPLEX', '-DUSE_ALPHA']},
+            '3m_iucopyr': {'addl': ['-DDOUBLE', '-DCOMPLEX', '-UUSE_ALPHA', '-DREAL_ONLY']},
+            '3m_oucopyr': {'addl': ['-DDOUBLE', '-DCOMPLEX', '-DUSE_ALPHA', '-DREAL_ONLY']},
+            '3m_iucopyi': {'addl': ['-DDOUBLE', '-DCOMPLEX', '-UUSE_ALPHA', '-DIMAGE_ONLY']},
+            '3m_oucopyi': {'addl': ['-DDOUBLE', '-DCOMPLEX', '-DUSE_ALPHA', '-DIMAGE_ONLY']},
+            '3m_olcopyb': {'addl': ['-DDOUBLE', '-DCOMPLEX', '-DUSE_ALPHA']},
+            '3m_ilcopyr': {'addl': ['-DDOUBLE', '-DCOMPLEX', '-UUSE_ALPHA', '-DREAL_ONLY']},
+            '3m_olcopyr': {'addl': ['-DDOUBLE', '-DCOMPLEX', '-DUSE_ALPHA', '-DREAL_ONLY']},
+            '3m_ilcopyi': {'addl': ['-DDOUBLE', '-DCOMPLEX', '-UUSE_ALPHA', '-DIMAGE_ONLY']},
+            '3m_olcopyi': {'addl': ['-DDOUBLE', '-DCOMPLEX', '-DUSE_ALPHA', '-DIMAGE_ONLY']},
```

Comment:
Does the Makefile/CMakefile list these all out separately or is there some way to generate this?

Reply:
The `Makefiles` do define each of these separately, but they can be cleaned up a bit (`CMake` has some functions for this, we use a cascade of dictionaries in `meson`). e.g. the `DOUBLE, COMPLEX` bits are based on the "mode" (`c`, `z`) and there's a pattern to be taken for the `{i,o}{u,l}copy_{b,r,i}` bits as well.

Theres's a bit of a readability issue though TBH, e.g. here's a bit from the `make` log:

```
cc -O2 -DSMALL_MATRIX_OPT -DMS_ABI -DMAX_STACK_ALLOC=2048 -Wall -m64 -DF_INTERFACE_GFORT -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=12 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.27.dev\" -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mavx2 -UASMNAME -UASMFNAME -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME=chemm3m_iucopyb -DASMFNAME=chemm3m_iucopyb_ -DNAME=chemm3m_iucopyb_ -DCNAME=chemm3m_iucopyb -DCHAR_NAME=\"chemm3m_iucopyb_\" -DCHAR_CNAME=\"chemm3m_iucopyb\" -DNO_AFFINITY -I.. -UDOUBLE  -DCOMPLEX -Wno-uninitialized -c -UDOUBLE -DCOMPLEX -UUSE_ALPHA generic/zhemm3m_ucopy_8.c -o chemm3m_iucopyb.obj
```

Where `-UDOUBLE -DCOMPLEX` gets repeated (and some other bits in the rest) because they're multiply defined (since its easier to see the whole command in context instead of parsing the build log).

The `meson` dictionary cascade and the CMake functions both require the reader to go looking instead of seeing a simpler symbol -> compilation map.